rpcs3/Emu/Cell/SPULLVMRecompiler.cpp

```diff
@@ -1672,35 +1676,120 @@ class spu_llvm_recompiler : public spu_recompiler_base, public cpu_translator
 			}
 
 			// Get actual pc corresponding to the found beginning of the data
-			llvm::Value* starta_pc = m_ir->CreateAnd(get_pc(starta), 0x3fffc);
+			llvm::Value* starta_pc = m_ir->CreateAnd(get_pc(starta), SPU_LS_SIZE - 4);
 			llvm::Value* data_addr = m_ir->CreateGEP(get_type<u8>(), m_lsptr, starta_pc);
 
 			llvm::Value* acc = nullptr;
 
-			// Use 512bit xorsum to verify integrity if size is atleast 512b * 3
-			// This code uses a 512bit vector for all hardware to ensure behavior matches.
+			// Exempt interrupt handler
+			bool has_sync_or_channel_sync = func.entry_point == 0;
+
+			for (u32 i = 0; !has_sync_or_channel_sync && i < func.data.size(); i++)
+			{
+				const spu_opcode_t op{std::bit_cast<be_t<u32>>(func.data[i])};
+
+				switch (g_spu_itype.decode(op.opcode))
+				{
+				case spu_itype::RDCH:
+				{
+					if (op.ra == MFC_RdTagStat)
+					{
+						//has_sync_or_channel_sync = true;
+					}
+
+					break;
+				}
+				case spu_itype::RCHCNT:
+				{
+					if (op.ra == MFC_RdTagStat || op.ra == SPU_WrOutIntrMbox || op.ra == SPU_WrOutMbox)
+					{
+						// Actually this is fine and is quite common
+						break;
+					}
+
+					has_sync_or_channel_sync = true;
+					break;
+				}
+				case spu_itype::SYNC:
+				case spu_itype::IRET:
+				case spu_itype::BISLED:
+				{
+					// Disable optimization if there is a SYNC or interrupt instruction
+					has_sync_or_channel_sync = true;
+					break;
+				}
+				case spu_itype::BI:
+				{
+					// Interrupt enabling instruction
+					has_sync_or_channel_sync = !!op.e;
+					break;
+				}
+				case spu_itype::STQR:
+				{
+					// Self-modifying code is likely
+					has_sync_or_channel_sync = (op.si16 >= 0 && op.i16 < func.data.size() - i) || (op.si16 < 0 && (0 - op.si16) + 0u < i);
+					break;
+				}
+				//case spu_itype::STQA:
+				default:
+				{
+					break;
+				}
+				}
+			}
+
+			// Use 512bit xorsum to verify integrity if size is atleast (stride * 2)
 			// The xorsum path is still faster even on narrow hardware.
-			if ((end - starta) >= 192 && !g_cfg.core.precise_spu_verification)
+			if (!has_sync_or_channel_sync && (end - starta) >= std::max<u32>(stride, 32) * 2 && !g_cfg.core.precise_spu_verification)
```

Comment:
I don't think it's a good idea to use the xorsum for such small functions. For 128 and 256b machines, the precise method would be faster, and for 512b machines, it would just end up being the same speed.

Remember, comparing 512b is not super cheap on those machines, it's only faster because with larger functions we skip out on loading and comparing even more data.

Reply:
I'm not using 512b for these machines. If it's an optimization of small probablity I say let's go all the way and use the native comparison. Even if it mismatches.

Reply:
I pushed an update to review, although it is still broken,

Reply:
Using 512b for the older machines wasn't an accident, it was both to ensure compatibility, and to ensure that the code operated at L1 cache speed, as common machines do 2x256b loads/clock, so we need more than dependency chain for max speed. It wasn't an issue for 512b since avx512 has vpternlog.